File: walkthroughs/python/anatomy_of_a_parquet_file/with_a_library.py
# The same facts from pyarrow, which reads the trailer and parses the footer for you.
import pyarrow.parquet as pq

metadata = pq.read_metadata("fixtures/tiny.parquet")
print("footer length:", metadata.serialized_size)
print("rows:", metadata.num_rows, "in", metadata.num_row_groups, "row group")
for i in range(metadata.num_columns):
    chunk = metadata.row_group(0).column(i)
    # A chunk starts with its dictionary page when it has one, and its first data page otherwise.
    start = chunk.dictionary_page_offset or chunk.data_page_offset
    print(f"column chunk {chunk.path_in_schema}: bytes {start} to {start + chunk.total_compressed_size}")

File: walkthroughs/libraries/src/bin/with_a_library.rs
//! The same facts from the `parquet` crate, which reads the trailer and parses the footer for you.

use parquet::file::metadata::{FooterTail, ParquetMetaDataReader};

fn main() {
    let file =
        std::fs::File::open("fixtures/tiny.parquet").expect("run this from the repository's root");
    let metadata = ParquetMetaDataReader::new()
        .parse_and_finish(&file)
        .expect("a Parquet file");

    // The crate checks the trailer as it parses; it also decodes one on its own, from eight bytes.
    let data = std::fs::read("fixtures/tiny.parquet").unwrap();
    let tail = FooterTail::try_from(&data[data.len() - 8..]).expect("a Parquet trailer");
    println!("footer length: {}", tail.metadata_length());
    println!(
        "rows: {} in {} row group",
        metadata.file_metadata().num_rows(),
        metadata.num_row_groups()
    );
    for chunk in metadata.row_group(0).columns() {
        // A chunk starts with its dictionary page when it has one, and its first data page otherwise.
        let (start, length) = chunk.byte_range();
        println!(
            "column chunk {}: bytes {start} to {}",
            chunk.column_path().string(),
            start + length
        );
    }
}
